### PR TITLE
Introduce `RenateLightningModule`

### PR DESCRIPTION
**Pull request by wistuba**
Split Learner into two classes. RenateLightningModule contains the basic logic while Learner adds all CL-related stuff such as val_buffer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/simple_classifier_cifar10/renate_config.py [hunk 1]
```diff
@@ -36,7 +37,7 @@ def data_module_fn(data_path: str, chunk_id: int, seed: int = defaults.SEED) ->
     )
     class_incremental_scenario = ClassIncrementalScenario(
         data_module=data_module,
-        class_groupings=[[0, 1, 2, 3, 4], [5, 6, 7, 8, 9]],
+        class_groupings=((0, 1, 2, 3, 4), (5, 6, 7, 8, 9)),
```
Comment (prabhuteja12) [inline examples/simple_classifier_cifar10/renate_config.py:40]: Why has this been changed?

File: src/renate/benchmark/experiment_config.py [hunk 2]
```diff
@@ -317,7 +318,7 @@ def train_transform(dataset_name: str) -> Optional[transforms.Compose]:
     raise ValueError(f"Unknown dataset `{dataset_name}`.")
 
 
-def test_transform(dataset_name: str) -> Optional[transforms.Normalize]:
+def test_transform(dataset_name: str) -> Optional[Union[transforms.Normalize, transforms.Compose]]:
```
Comment (prabhuteja12) [inline src/renate/benchmark/experiment_config.py:321]: Can this be made a callable? This is very specific to existing transformations.

File: examples/nlp_finetuning/start.py [hunk 3]
```diff
@@ -25,8 +25,8 @@
 
     run_training_job(
         config_space=config_space,
-        mode="max",
-        metric="val_accuracy",
+        mode="min",
```
Comment (prabhuteja12) [inline examples/nlp_finetuning/start.py:28]: Does this make a difference?

File: src/renate/updaters/learner.py [hunk 4]
```diff
@@ -76,41 +64,35 @@ def __init__(
         self._learning_rate_scheduler = learning_rate_scheduler
         self._learning_rate_scheduler_interval = learning_rate_scheduler_interval
         self._batch_size = batch_size
-        self._train_transform = train_transform
-        self._train_target_transform = train_target_transform
-        self._test_transform = test_transform
-        self._test_target_transform = test_target_transform
         self._seed = seed
         self._task_id: str = defaults.TASK_ID
+        self._train_dataset = None
```
Comment (prabhuteja12) [inline src/renate/updaters/learner.py:69]: no typing :)

File: src/renate/updaters/learner.py [hunk 5]
```diff
@@ -224,11 +182,17 @@ def forward(self, inputs: NestedTensors, task_id: Optional[str] = None) -> torch
             task_id = self._task_id
         return self._model(inputs, task_id=task_id)
 
+    def training_step_unpack_batch(
+        self, batch: Tuple[NestedTensors, torch.Tensor]
+    ) -> Tuple[NestedTensors, Any]:
+        inputs, targets = batch
```
Comment (prabhuteja12) [inline src/renate/updaters/learner.py:188]: This choice of what a batch comprises is specific to HF vs non-HF datasets? So, why is this specified in the base class?

File: src/renate/updaters/learner.py [hunk 6]
```diff
@@ -253,9 +217,15 @@ def training_epoch_end(self, outputs: List[Union[Tensor, Dict[str, Any]]]) -> No
         if not self.val_enabled:
             self._log_metrics()
 
+    def validation_step_unpack_batch(
+        self, batch: Tuple[NestedTensors, torch.Tensor]
+    ) -> Tuple[NestedTensors, Any]:
```
Comment (prabhuteja12) [inline src/renate/updaters/learner.py:222]: same as above.

File: examples/simple_classifier_cifar10/renate_config.py (hunk 1 above)
Comment (wistuba) [reply examples/simple_classifier_cifar10/renate_config.py:40]: To match typing.

File: examples/nlp_finetuning/start.py (hunk 3 above)
Comment (wistuba) [reply examples/nlp_finetuning/start.py:28]: yes. accuracy is no longer computed by default and since it was not added to the renate config, using it wouldn't be possible. we can keep it but then we need to add the metric.

File: src/renate/benchmark/scenarios.py [hunk 7]
```diff
@@ -56,7 +59,10 @@ def prepare_data(self) -> None:
     @abc.abstractmethod
```
Comment (prabhuteja12) [inline src/renate/benchmark/scenarios.py:59]: What does an `abc` with code do?

File: src/renate/benchmark/experiment_config.py (hunk 2 above)
Comment (prabhuteja12) [reply src/renate/benchmark/experiment_config.py:321]: Why is this also `Optional`?

**Review by prabhuteja12 — CHANGES_REQUESTED** [review]
Few small changes. I agree with the overall structure of what is being accomplished.

File: src/renate/benchmark/experiment_config.py (hunk 2 above)
Comment (wistuba) [reply src/renate/benchmark/experiment_config.py:321]: because it can return `None`

File: src/renate/benchmark/scenarios.py (hunk 7 above)
Comment (wistuba) [reply src/renate/benchmark/scenarios.py:59]: Forces the implementation of the function:

```
import abc

class A(abc.ABC):
    @abc.abstractmethod
    def foo(self):
        print("parent")
        
        
class B(A):
    def foo(self):
        super().foo()
        
class C(A):
    pass

A().foo()  # Fails
B().foo()  # Works
C().foo()  # Fails
```

**Review by prabhuteja12 — APPROVED** [review]
